### PR TITLE
Fix/max safe volume and roundup

### PR DESCRIPTION
**Pull request by adhusson**
- MAX_SAFE_VOLUME was off by x2
- inboundFromOutboundUp and outboundFromInboundUp now correctly round up when exp of floating-point price is > 255

### CODE REVIEW

File: lib/BitLib.sol [hunk 1]
```diff
@@ -2,21 +2,33 @@
...
+        // isolate lsb
+        x := and(x, add(not(x), 1))
+
+        // 6th bit
+        c := or(c,shl(5, lt(0xffffffff, x)))
+
+        // debruijn lookup
+        c := or(c, byte(shr(251, mul(shr(c, x), shl(224, 0x077cb531))), 
+            0x00011c021d0e18031e16140f191104081f1b0d17151310071a0c12060b050a09))
+      }
```
Comment (espendk) [inline lib/BitLib.sol:27]: I haven't checked this, let me know if I should.
Is it an optimized version of the previous implementation for 64 bits?

File: lib/LogPriceLib.sol [hunk 2]
```diff
@@ -52,11 +53,28 @@ library LogPriceLib {
...
+}
 
+// Return a/2**e rounded up
+function divExpUp(uint a, uint e) pure returns (uint) {
+  unchecked {
+    uint carry;
+    /* 
+    Would like to write
 
+      carry = (e<<1 == 0 || prod % (e<<1) != 0) ? 1 : 0
```
Comment (espendk) [inline lib/LogPriceLib.sol]: `prod` -> `a`
```suggestion
      carry = (e<<1 == 0 || a % (e<<1) != 0) ? 1 : 0
```

Comment (espendk) [reply lib/LogPriceLib.sol]: Why left-shift when checking for zero?
Can't we just do:

```
carry = (e == 0 || a % (e<<1) != 0) ? 1 : 0
```

Comment (espendk) [reply lib/LogPriceLib.sol]: In the event that `e == 0  =>  2**e == 1` shouldn't `carry == 1` iff `a % 2 == 1`?

I think the current formula adds 1 when `e == 0`  and  `a % 2 == 0`, but I don't think it should?

Comment (espendk) [reply lib/LogPriceLib.sol]: Oh, should the expression `e<<1`have been `1<<e`?
That would correspond to $2^e$ which would make sense here :-)

(the fact that Yul's `shl(a,b)` corresponds to `b<<a` is confusing)

File: lib/TickLib.sol [hunk 3]
```diff
@@ -62,110 +63,109 @@ library LeafLib {
...
-        } else {
-          offerId = Leaf.unwrap(leaf) << (OFFER_BITS * 6) >> (OFFER_BITS * 7);
-          if (offerId != 0) {
-            ret = 3;
-          }
-        }
+    unchecked {
+      assembly("memory-safe") {
+        ret := shl(1,gt(0xffffffffffffffffffffffffffffffff,leaf))
+        ret := or(ret,gt(0xffffffffffffffff,shr(shl(7,iszero(ret)),leaf)))
```
Comment (espendk) [inline lib/TickLib.sol:161]: I have not checked this. Let me know if I should

File: lib/TickLib.sol [hunk 4]
```diff
@@ -174,33 +174,49 @@ library LeafLib {
...
-          (BitLib.ctz(Field.unwrap(level1)) |
-            (BitLib.ctz(Field.unwrap(level2)) |
-              uint((int(BitLib.ctz(Field.unwrap(level3)))-LEVEL3_SIZE/2) << LEVEL2_SIZE_BITS)) 
+      LocalPacked local;
+      local = local.tickPosInLeaf(tickPosInLeaf).level0(level0).level1(level1).level2(level2).level3(level3);
+      return tickFromLocal(local);
+    }
+  }
+
+  function tickFromLocal(LocalPacked local) internal pure returns (Tick) {
```
Comment (espendk) [inline lib/TickLib.sol]: These return the *best* tick on that branch, right? If so, would be good if the name reflected this, eg, `bestTickFromBranch`.

File: test/core/Leaf.t.sol [hunk 5]
```diff
@@ -286,12 +293,18 @@ contract TickTest is Test {
...
     Field level2 = Field.wrap(bound(_level2, 1, uint(LEVEL2_SIZE) - 1));
     Field level3 = Field.wrap(bound(_level3, 1, uint(LEVEL3_SIZE) - 1));
-    Tick tick = TickLib.tickFromBranch(tickPosInLeaf, level0, level1, level2, level3);
+    MgvStructs.LocalPacked local;
+    local = local.tickPosInLeaf(tickPosInLeaf);
+    local = local.level0(level0);
+    local = local.level1(level1);
+    local = local.level2(level2);
+    local = local.level3(level3);
+    Tick tick = TickLib.tickFromLocal(local);
```
Comment (espendk) [inline test/core/Leaf.t.sol:302]: Test name still refers to `ticksFromBranch`: `test_tickFromBranch_matches_positions_accessor`

File: test/core/Leaf.t.sol [hunk 6]
```diff
@@ -351,7 +364,11 @@ contract FieldTest is Test {
     for (; i < 256; i++) {
       if (uint(b >> i) % 2 == 1) break;
     }
-    assertFirstOnePosition(uint(b), i);
+    assertFirstOnePosition(uint(b), i > MAX_LEVEL_SIZE ? MAX_LEVEL_SIZE : i);
+  }
+
+  function test_regress() public {
+    test_firstOnePosition_auto(14530771982722103506997891882992560810543470234613570268015259683728373317632);
```
Comment (espendk) [inline test/core/Leaf.t.sol:371]: Should this stay here?

File: lib/LogPriceLib.sol [hunk 7]
```diff
@@ -52,11 +53,28 @@ library LogPriceLib {
...
 
+    but solidity reverts on a%0 even inside unchecked blocks
+    */
+    assembly("memory-safe") {
+      let den := shl(e,1)
+      carry := gt(or(mod(a,den),iszero(den)),0)
```
Comment (espendk) [inline lib/LogPriceLib.sol]: When can `iszero(den) == iszero(shl(e,1))` be true? When `e > 255`?

Comment (espendk) [reply lib/LogPriceLib.sol]: If so, shouldn't there only be a carry if `a > 0`?

File: test/core/Leaf.t.sol [hunk 8]
```diff
@@ -463,4 +480,53 @@ contract FieldTest is Test {
...
+    //inboundFromOutboundUp
+    (sig, exp) = LogPriceConversionLib.nonNormalizedPriceFromLogPrice(logPrice);
+    assertEq(LogPriceLib.inboundFromOutboundUp(logPrice, amt), divExpUp_spec(sig * amt, exp));
+
+    //outboundFromInboundUp
+    (sig, exp) = LogPriceConversionLib.nonNormalizedPriceFromLogPrice(-logPrice);
+    assertEq(LogPriceLib.outboundFromInboundUp(logPrice, amt), divExpUp_spec(sig * amt, exp));
+  }
+
+  function test_outuboundFromInboundUp(int logPrice, uint amt) public {
```
Comment (espendk) [inline test/core/Leaf.t.sol]:
```suggestion
  function test_outboundFromInboundUp(int logPrice, uint amt) public {
```

File: test/core/Leaf.t.sol [hunk 9]
```diff
@@ -463,4 +480,53 @@ contract FieldTest is Test {
...
+    amt = bound(amt, 0, MAX_SAFE_VOLUME);
+    logPrice = bound(logPrice, MIN_LOG_PRICE, MAX_LOG_PRICE);
+
+    (uint sig, uint exp) = LogPriceConversionLib.nonNormalizedPriceFromLogPrice(logPrice);
+    assertEq(LogPriceLib.inboundFromOutboundUp(logPrice, amt), divExpUp_spec(sig * amt, exp));
```
Comment (espendk) [inline test/core/Leaf.t.sol]: Test name indicates this should test `outboundFromInboundUp`?

File: test/core/Leaf.t.sol (hunk 8 above)
Comment (espendk) [reply test/core/Leaf.t.sol]: Does the previous test cover the same?

File: test/core/gas/MarketOrderOtherOfferList.t.sol [hunk 10]
```diff
@@ -234,6 +234,8 @@ contract ExternalMarketOrderOtherOfferList_WithMultipleOffersAtManyTicks is Tick
     mgv.marketOrderByLogPrice(_olKey, _logPrice, 2 ** 96, false);
     gas_();
     (, MgvStructs.LocalPacked local) = mgv.config(_olKey);
-    assertLt(_logPrice, LogPriceLib.fromTick(local.bestTick(), _olKey.tickScale));
+    if (!local.level3().isEmpty()) {
```
Comment (espendk) [inline test/core/gas/MarketOrderOtherOfferList.t.sol:237]: How come this is necessary now? Does it mean some assumption in the tests no longer holds?

**Review by espendk — COMMENTED** [review]
Gnarly stuff 😅 And more than expected for this fix. I added some comments and questions.

I did not review the optimizations, let me know if I should. Would appreciate an intro in that case :-)

File: lib/BitLib.sol (hunk 1 above)
Comment (adhusson) [reply lib/BitLib.sol:27]: yes it's an optimized version -- we can check this later :)

File: lib/TickLib.sol (hunk 3 above)
Comment (adhusson) [reply lib/TickLib.sol:161]: see #544

File: lib/TickLib.sol (hunk 4 above)
Comment (adhusson) [reply lib/TickLib.sol]: see #544

File: test/core/Leaf.t.sol (hunk 5 above)
Comment (adhusson) [reply test/core/Leaf.t.sol:302]: see #544

File: test/core/Leaf.t.sol (hunk 6 above)
Comment (adhusson) [reply test/core/Leaf.t.sol:371]: see #544

File: test/core/gas/MarketOrderOtherOfferList.t.sol (hunk 10 above)
Comment (adhusson) [reply test/core/gas/MarketOrderOtherOfferList.t.sol:237]: see #544

File: lib/LogPriceLib.sol (hunk 2 above)
Comment (adhusson) [reply lib/LogPriceLib.sol]: sorry this was stale, i'm pushing a cleaner version that addresses your comments

File: test/core/Leaf.t.sol (hunk 8 above)
Comment (adhusson) [reply test/core/Leaf.t.sol]: test removed

File: test/core/Leaf.t.sol (hunk 9 above)
Comment (adhusson) [reply test/core/Leaf.t.sol]: test removed

File: lib/LogPriceLib.sol (hunk 7 above)
Comment (adhusson) [reply lib/LogPriceLib.sol]: pushed a fix :)

**Comment by adhusson** [comment]
closed for #553

File: test/core/gas/MarketOrderOtherOfferList.t.sol (hunk 10 above)
Comment (espendk) [reply test/core/gas/MarketOrderOtherOfferList.t.sol:237]: #554 I believe

**Pull request closed without merging**